### ISSUE TITLE
Bit identical copy of an unencrypted storage to an encrypted storage waives encryption?

### ISSUE THREAD

**Issue by ei-ke** [gilbertchen/duplicacy#284]
First I have to say thank you to Gang for his extensive support I received by email the past days.
I brought this issue to his attention but I think I haven't described the issue, if it is an issue, as detailed as I will do here.

Grab a publicly available sample set of files and set up the workbench
```
mkdir ~/duplicacy_test && cd ~/duplicacy_test && wget https://github.com/torvalds/linux/archive/v4.14.tar.gz && tar -zxvf v4.14.tar.gz && rm  v4.14.tar.gz && mkdir storage_1 && mkdir storage_2 && cd linux-4.14
```

## **First setup**
**deafult storage unencrypted > make a backup > add an encrypted bit-identical second storage > copy backup**

Initialize the unencrypted default storage
`duplicacy init linux-src ~/duplicacy_test/storage_1`
>```
>/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_1 with id >linux-src
>```

Make the first backup
`duplicacy backup -stats > ~/duplicacy_test/storage_1_v1.log`
> [storage_1_v1.log](https://github.com/gilbertchen/duplicacy/files/1523121/storage_1_v1.log)

Add the second storage which will be an bit-identical encrypted one
`duplicacy add -encrypt -copy default -bit-identical storage_2 linux-src ~/duplicacy_test/storage_2`

> ```
> Enter storage password for /home/user/duplicacy_test/storage_2:11111111
> 	Re-enter storage password:11111111
> 	/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
> ```

Wirte the encryption password to the configuration
`duplicacy set -storage storage_2 -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Copy backup from the default unencrypted storage to the encrypted bit-identical second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v1.log`
>[storage_2_v1.log](https://github.com/gilbertchen/duplicacy/files/1523152/storage_2_v1.log)

Let's use diff to see if the data is actually encrypted - `md5sum` et al. will work as well
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2`
> ```
> Dateien /home/user/duplicacy_test/storage_1/config und /home/user/duplicacy_test/storage_2/config sind verschieden.
> ```
I'm not translating this, but the only difference found was for the config file.

Conlusion: Something here went wrong or, as the config file changed and is now binary and not a text file, all relevant stuff that needs to be encrypted lives in the config file.

## **Second setup**
**deafult storage unencrypted > make a backup > add an unencrypted bit-identical second storage > copy backup**

Removing storage_2 from this setup
`perl -0777 -pi -e 's/\,\s{5}\{((.|\n)*?)\n\s{4}\}//g' .duplicacy/preferences`
`rm -r ~/duplicacy_test/storage_2/*`

Add second unencrypted bit-identical storage
`duplicacy add -copy default -bit-identical storage_2 linux-src ~/duplicacy_test/storage_2`
>```
>/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
>```

Copy backup from the default unencrypted storage to the unencrypted second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v2.log`
> [storage_2_v2.log](https://github.com/gilbertchen/duplicacy/files/1523151/storage_2_v2.log)

Diff
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2`
> No output in this case means there are no differences
Conclusion: Maybe the magic is really in the config file...

## **Third setup**
**deafult storage unencrypted > make a backup > add an encrypted non-bit-identical second storage > copy backup**

Removing storage_2 from this setup
`perl -0777 -pi -e 's/\,\s{5}\{((.|\n)*?)\n\s{4}\}//g' .duplicacy/preferences`
`rm -r ~/duplicacy_test/storage_2/*`

Add second non-bit-identical encrypted storage
`duplicacy add -encrypt -copy default storage_2 linux-src ~/duplicacy_test/storage_2`
> ```
> Enter storage password for /home/user/duplicacy_test/storage_2:11111111
> Re-enter storage password:11111111
> /home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
> ```

Wirte the encryption password to the configuration
`duplicacy set -storage storage_2 -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Copy backup from the default unencrypted storage to the encrypted non-bit-identical second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v3.log`
>[storage_2_v3.log](https://github.com/gilbertchen/duplicacy/files/1523149/storage_2_v3.log)

Diff
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2 > ~/duplicacy_test/diff_v1.log`
>[diff_v1.log](https://github.com/gilbertchen/duplicacy/files/1523144/diff_v1.log)
Not sure how I should interpret this.

## **Fourth setup**
**deafult storage encrypted > make a backup > add an encrypted bit-identical second storage > copy backup**
Remove all data created by duplicacy
```
rm -r ~/duplicacy_test/storage_2/* && rm -r ~/duplicacy_test/storage_1/* && rm -r ~/duplicacy_test/linux-4.14/.duplicacy/
```
Initialize the encrypted default storage
`duplicacy init -encrypt linux-src ~/duplicacy_test/storage_1`

> ```
> Enter storage password for /home/eike/duplicacy_test/storage_1:11111111
> Re-enter storage password:11111111
> /home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_1 with id linux-src
> ```
Wirte the encryption password to the configuration
`duplicacy set -storage default -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```
Make the first backup
`duplicacy backup -stats > ~/duplicacy_test/storage_1_v2.log`
>[storage_1_v2.log](https://github.com/gilbertchen/duplicacy/files/1523171/storage_1_v2.log)

Add the second storage which will be an bit-identical encrypted one
`duplicacy add -encrypt -copy default -bit-identical storage_2 linux-src ~/duplicacy_test/storage_2`

> ```
> Enter storage password for /home/user/duplicacy_test/storage_2:11111111
> 	Re-enter storage password:11111111
> 	/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
> ```

Wirte the encryption password to the configuration
`duplicacy set -storage storage_2 -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Copy backup from the default encrypted storage to the encrypted bit-identical second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v4.log`
>[storage_2_v4.log](https://github.com/gilbertchen/duplicacy/files/1523176/storage_2_v4.log)

Diff
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2  > ~/duplicacy_test/diff_v2.log`
>[diff_v2.log](https://github.com/gilbertchen/duplicacy/files/1523179/diff_v2.log)
Same filenames but different file content.

## **Fifth setup**
**deafult storage encrypted > make a backup > add an encrypted non-bit-identical second storage > copy backup**

Removing storage_2 from this setup
`perl -0777 -pi -e 's/\,\s{5}\{((.|\n)*?)\n\s{4}\}//g' .duplicacy/preferences`
`rm -r ~/duplicacy_test/storage_2/*`

Add second non-bit-identical encrypted storage
`duplicacy add -encrypt -copy default storage_2 linux-src ~/duplicacy_test/storage_2`
> ```
> Enter storage password for /home/user/duplicacy_test/storage_2:11111111
> Re-enter storage password:11111111
> /home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
> ```

Wirte the encryption password to the configuration
`duplicacy set -storage storage_2 -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Copy backup from the default unencrypted storage to the encrypted non-bit-identical second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v5.log`
>[storage_2_v5.log](https://github.com/gilbertchen/duplicacy/files/1523190/storage_2_v5.log)

Diff
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2 >  ~/duplicacy_test/diff_v3.log`
>[diff_v3.log](https://github.com/gilbertchen/duplicacy/files/1523191/diff_v3.log)

Not sure how I should interpret this.

## **Sixth setup**
**deafult storage unencrypted > add an encrypted bit-identical second storage > make a backup > copy backup##
Remoe all data created by duplicacy
```
rm -r ~/duplicacy_test/storage_2/* && rm -r ~/duplicacy_test/storage_1/* && rm -r ~/duplicacy_test/linux-4.14/.duplicacy/
```
Initialize the encrypted default storage
`duplicacy init linux-src ~/duplicacy_test/storage_1`

>```
>/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_1 with id >linux-src
>```

Wirte the encryption password to the configuration
`duplicacy set -storage default -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Add the second storage which will be an bit-identical encrypted one
`duplicacy add -encrypt -copy default -bit-identical storage_2 linux-src ~/duplicacy_test/storage_2`
> ```
> Enter storage password for /home/user/duplicacy_test/storage_2:11111111
> 	Re-enter storage password:11111111
> 	/home/user/duplicacy_test/linux-4.14 will be backed up to /home/user/duplicacy_test/storage_2 with id linux-src
> ```

Wirte the encryption password to the configuration
`duplicacy set -storage storage_2 -key password -value 11111111`
>```
>New options for storage /home/user/duplicacy_test/storage_2 have been saved
>```

Make the first backup
`duplicacy backup -stats > ~/duplicacy_test/storage_1_v3.log`
>[storage_1_v3.log](https://github.com/gilbertchen/duplicacy/files/1523207/storage_1_v3.log)

Copy backup from the default encrypted storage to the encrypted bit-identical second storage
`duplicacy copy -from default -to storage_2 > ~/duplicacy_test/storage_2_v6.log`
>[storage_2_v6.log](https://github.com/gilbertchen/duplicacy/files/1523208/storage_2_v6.log)

Diff
`diff -rq ~/duplicacy_test/storage_1 ~/duplicacy_test/storage_2`
> ```
> Dateien /home/user/duplicacy_test/storage_1/config und /home/user/duplicacy_test/storage_2/config sind verschieden.
> ```
I'm not translating this, but the only difference found was for the config file.
Basically the same result as in setup one - so adding the encrypted bit-identical storage before the first backup is ever done doesn't change anything.

# **Final result**
I still believe there is a bug and that a bit-identical copy of an unencrypted storage currently reduces the encryption to absurdity as it seems to be non existent.

PS: I hope I've not messed up what I wrote above and that it can be reproduced.

**Comment by gilbertchen** [comment]
Thank you for such detailed report.  I think the only issue here is with the first setup where the `-bit-identical` option was used for an unencrypted source storage and an encrypted destination storage.  In this case, the new config file for the encrypted destination storage is an encrypted version of the standard unencrypted config file with default values, but you can easily replace it with the standard unencrypted one, so there is really no point in doing this.  The fix I think is to prohibit the use of `-bit-identical` when copying the config file from an unencrypted storage to an encrypted one.

**Comment by ei-ke** [comment]
Ok. 
I've seen the scheme how the encryption works for duplicacy. But maybe I just don't understand it or also don't understand the copy command. 

1) When copying the encrypted default storage bit-identical to the second encrypted destination with the exact same passphrase I would have expected that also the chunks will be the same when using diff which is not the case. Maybe you can tell me why that's the case. 

2) Further as per your comment,  if I understood it right, it is not only important that the config is encrypted but the chunks also need to be encrypted?
So than maybe for the first case just encrypt the chunks and everything will be ok as a solution. At least when copying bit-identical from encrypted to encrypted,  from what I see when using diff, the content of the chunks change which indicates they are reencrypted. 

3) So when the default storage is encrypted I can be sure that this includes the config an the chunks?

**Comment by gilbertchen** [comment]
> When copying the encrypted default storage bit-identical to the second encrypted destination with the exact same passphrase I would have expected that also the chunks will be the same when using diff which is not the case. Maybe you can tell me why that's the case.

In this case, the chunks will be re-encrypted with different salts, so the resulting files will be different.  However, if you simply copy the chunks over using rsync/rclone then they should work too.  This is one of the reasons you want bit-identical.

> Further as per your comment, if I understood it right, it is not only important that the config is encrypted but the chunks also need to be encrypted?  So than maybe for the first case just encrypt the chunks and everything will be ok as a solution. At least when copying bit-identical from encrypted to encrypted, from what I see when using diff, the content of the chunks change which indicates they are reencrypted.

The unencrypted `config` file has empty encryption keys which essentially disable the encryption.  The bit-identical option copies these empty encryption keys too so now the chunks become unencrypted (the `config` file is still encrypted by the storage password).

> So when the default storage is encrypted I can be sure that this includes the config an the chunks?

This is correct.

**Comment by ei-ke** [comment]
Thank you for being patient with me. Now a blockhead like me in this case also got it :)

**Comment by sergeevabc** [comment]
@ei-ke, has your question been answered? If you’re satisfied, please, do close the issue.

**Comment by ei-ke** [comment]
@sergeevabc Thanks for the hint. I thought gilbertchen would close this whenever he likes to close it :)
